rfc-drafts/yield_safe_lint.md

```diff
@@ -0,0 +1,89 @@
+# RFC: yield safe lint
+
+# Summary
+
+Introduce a `#[yield_unsafe]` lint in the compiler that will warn the user when they unsafely hold an unsafe yield struct across a yield boundary. 
```

Comment:
> [...] the compiler that will warn the user when they unsafely hold an unsafe yield struct across a yield boundary

It seems the term "unsafely" doesn't seem to refer to memory safety here? The terms "safe" and "unsafe" have very specific meanings in Rust. I would strongly caution against repurposing the terminology of `safe` and `unsafe` for anything unrelated to its current use.

Teaching people about unsafe Rust is pretty hard already, and it's often misunderstood. I think it would be bad for the language if the meaning of "unsafe" within Rust became dependent on context.

Instead I would suggest using terminology such as:
- `#[no_yield]` / "can be yielded" / "cannot be yielded"
- `#[disallow_yield]` / "allowed to be yielded" / "not allowed to be yielded"
- `#[deny_yield]` / "allowed to be yielded" / "not allowed to be yielded"

Reply:
(actually, this seems mentioned in "Unresolved questions" section)

There is a case where "safe" is used as a safe API in places where it doesn't have to do with memory safety (`std::panic::UnwindSafe`), but it seems "unsafe" doesn't.

> * `#[deny_yield]` / "allowed to be yielded" / "not allowed to be yielded"

It seems odd that the lint name warned by default is "deny_*".

Reply:
> There is a case where "safe" is used as a safe API in places where it doesn't have to do with memory safety (std::panic::UnwindSafe), but it seems "unsafe" doesn't.

I don't believe that's *quite* true. From the explainer in the docs of what "unwind safety" is:

> It is possible, however, for logical invariants to be broken in Rust, which can end up causing behavioral bugs. Another key aspect of unwind safety in Rust is that, in the absence of unsafe code, a panic cannot lead to memory unsafety.

Which means that if a type is `!UnwindSafe`, it's not memory safe to unwind and the code won't compile. This seems quite different than what the `#[yield_unsafe]` lint introduces?

Reply:
What I wanted to say is that *the term "safe" is used in APIs that you can never violate memory safety without other unsafe code, even if implemented incorrectly*.

https://github.com/rust-lang/rfcs/blob/master/text/1236-stabilize-catch-panic.md#background-what-is-exception-safety-in-rust
> Despite these methods to subvert the mitigations placed by default in Rust, a key part of exception safety in Rust is that **safe code can never lead to memory unsafety**, regardless of whether it panics or not. Memory unsafety triggered as part of a panic can always be traced back to an unsafe block.

Reply:
Another point on naming: "yield" isn't a documented term in Rust today. While I would very much like to use terminology that's future-proof for when generators stabilize, it seems like it would be quite confusing in the meantime.

Maybe we could use "async" terminology for now, and simply add "yield" terminology as a synonym in the future. The RFC should discuss future options and potential drawbacks, if any.

Reply:
Yeah, I am not sold on unsafe it was just the best way I could think of explaining it to people that were familiar with async concepts. I would love to hear more proposals I am very open to naming.

@tmandry yeah, that is a really good point. We should probably use await then? But that also limits us when we do get generators out (which idk when that will happen) so not a huge deal we can duplicate the macro.

- `#[deny_hold_await]`
- `#[disallow_across_await]` this seems promising and similar to what @yoshuawuyts suggested.